File: rust/src/executors/sqlx/report.rs
use crate::model::domain::Period;
use sqlx::{Pool, Postgres};

pub async fn report(pool: &Pool<Postgres>, start_period: Period, end_period: Period) {
    let start_period_yp = start_period.year_period();
    let end_period_yp = end_period.year_period();

    let _start_balance = sqlx::query!(
        r#"select account_id, sum(amount) as amount 
                 from fin_ledger_items
                 where period < $1
                 group by account_id
               "#,
        start_period_yp,
    )
    .fetch_all(pool)
    .await
    .expect("Unable to retrieve ledger items");

    let _turnaround = sqlx::query!(
        r#"select account_id, sum(debt) as debt, sum(credit) as credit 
                 from fin_ledger_items
                 where period >= $1 and period <= $2
                 group by account_id
               "#,
        start_period_yp,
        end_period_yp,
    )
    .fetch_all(pool)
    .await
    .expect("Unable to retrieve ledger items");
}

File: golang/executors/gorm_ex/report.go
package gorm_ex

import (
	"bench-pg-go/executors/gorm_ex/model"
	"bench-pg-go/model/domain"
	"gorm.io/gorm"
)

func report(db *gorm.DB, startPeriod, endPeriod domain.Period) {
	startPeriodYp := startPeriod.YearPeriod()
	endPeriodYp := endPeriod.YearPeriod()

	var balance map[string]interface{}

	res := db.Model(&model.LedgerItemEntity{}).
		Select("account_id", "sum(amount)").
		Where("period < ? ", startPeriodYp).
		Group("account_id").
		Scan(&balance)
	if res.Error != nil {
		panic(res.Error)
	}

	var turnaround map[string]interface{}
	res = db.Model(&model.LedgerItemEntity{}).
		Select("account_id", "sum(debt) as debt", "sum(credit) as credit").
		Where("period >= ? and period <= ? ", startPeriodYp, endPeriodYp).
		Group("account_id").
		Scan(&turnaround)
	if res.Error != nil {
		panic(res.Error)
	}
}
